applications/MeshingApplication/custom_utilities/mmg_utilities.cpp

```diff
@@ -0,0 +1,2458 @@
+// KRATOS  __  __ _____ ____  _   _ ___ _   _  ____
+//        |  \/  | ____/ ___|| | | |_ _| \ | |/ ___|
+//        | |\/| |  _| \___ \| |_| || ||  \| | |  _
+//        | |  | | |___ ___) |  _  || || |\  | |_| |
+//        |_|  |_|_____|____/|_| |_|___|_| \_|\____| APPLICATION
+//
+//  License:		 BSD License
+//                       license: MeshingApplication/license.txt
+//
+//  Main authors:    Vicente Mataix Ferrandiz
+//
+
+// System includes
+
+// External includes
+/* The includes related with the MMG library */
+#if !defined(MMG_INCLUDES)
+#define MMG_INCLUDES
+#include "mmg/libmmg.h"
+#include "mmg/mmg2d/libmmg2d.h"
+#include "mmg/mmg3d/libmmg3d.h"
+#include "mmg/mmgs/libmmgs.h"
+#endif /* MMG_INCLUDES defined */
+
+// Project includes
+#include "custom_utilities/mmg_utilities.h"
+
+// NOTE: The following contains the license of the MMG library
+/* =============================================================================
+**  Copyright (c) Bx INP/Inria/UBordeaux/UPMC, 2004- .
+**
+**  mmg is free software: you can redistribute it and/or modify it
+**  under the terms of the GNU Lesser General Public License as published
+**  by the Free Software Foundation, either version 3 of the License, or
+**  (at your option) any later version.
+**
+**  mmg is distributed in the hope that it will be useful, but WITHOUT
+**  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+**  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
+**  License for more details.
+**
+**  You should have received a copy of the GNU Lesser General Public
+**  License and of the GNU General Public License along with mmg (in
+**  files COPYING.LESSER and COPYING). If not, see
+**  <http://www.gnu.org/licenses/>. Please read their terms carefully and
+**  use this copy of the mmg distribution only if you accept them.
+** =============================================================================
+*/
+
+namespace Kratos
+{
+// The member variables related with the MMG library
+MMG5_pMesh mMmgMesh; /// The mesh data from MMG
+MMG5_pSol  mMmgSol;  /// The metric variable for MMG
+// MMG5_pSol  mmgDisp; /// The displacement variable for MMG
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<MMGLibray TMMGLibray>
+void MmgUtilities<TMMGLibray>::PrintMmgMeshInfo(
+    SizeType& rNumberOfNodes,
+    array_1d<SizeType, 2>& rNumberOfConditions,
+    array_1d<SizeType, 2>& rNumberOfElements,
+    const IndexType EchoLevel
+    )
+{
+    rNumberOfNodes = mMmgMesh->np;
+    if (TMMGLibray == MMGLibray::MMG2D) { // 2D
```

Comment:
This function does too many things and should probably be split in two, i.e., one to get the parameters and another to do the printing.

Second, you could improve readability, reduce templates, simplify testing and reduce function size if you split this if-statement into an overload function set with parameter types to indicate if it is MMG2D, 3D, etc. Overloading still gets resolved at compile time so you don't necessarily loose performance.

Reply:
A simple example please?

Reply:
See https://github.com/KratosMultiphysics/Kratos/compare/meshing/mmg_io...meshing/mmg_io-patch-overload-set?expand=1

I stopped after refactoring `PrintAndGetMmgMeshInfo`, but I think it is enough to get the idea. Again, not blocking.

```
// mmg_utilities.h
struct MeshInfo2D
{
    SizeType NumberOfNodes;
    SizeType NumberOfEdges;
    SizeType NumberOfTriangles;
};
MeshInfo2D GetMmgMeshInfo2D();
void PrintMmgMeshInfo(const MeshInfo2D& rMeshInfo, IndexType EchoLevel = 0);
...
```
is what I mean by self-documenting.

Reply:
I prefer not do it that way, that mesh info is used in several places. And the thing that reduces the most the number of explicit templates is the current one